src/main/java/org/opentripplanner/middleware/cdp/ConnectedDataManager.java

```diff
@@ -0,0 +1,216 @@
+     */
+    public static String getAnonymizedTripHistory(Date start, Date end) {
+        return JsonUtils.toJson(
+            new TripHistory(
+                TripRequest.getAnonymizedTripRequests(start, end),
+                TripSummary.getAnonymizedTripSummaries(start, end)
+            )
+        );
+    }
+
+    /**
+     * Add to the upload list any dates between now and the last upload date. This will cover a new day once
+     * passed midnight and any days missed due to downtime.
+     */
+    public static void stageUploadDays() {
+        Date now = new Date();
+        TripHistoryUpload latest = TripHistoryUpload.getLatest();
+        if (latest == null) {
+            // No data held, add the previous day as the first day to be uploaded.
+            Persistence.tripHistoryUploads.create(
+                new TripHistoryUpload(
+                    getStartOfDay(
+                        getDateMinusNumberOfDays(now, 1)
+                    )
+                ));
+        } else {
+            Set<LocalDate> betweenDays = DateTimeUtils.getDatesBetween(
+                getDatePlusNumberOfDays(latest.uploadDate,1),
+                now
+            );
+            betweenDays.forEach(day -> {
+                Persistence.tripHistoryUploads.create(new TripHistoryUpload(getStartOfDay(day)));
+            });
+        }
+    }
+
+    /**
+     * Process incomplete upload dates. This will be uploads which are flagged as 'pending'. If the upload date is
+     * compiled and uploaded successfully, it is flagged as 'complete'.
+     */
+    public static void processTripHistory() {
+        List<TripHistoryUpload> incompleteUploads = getIncompleteUploads();
+        incompleteUploads.forEach(tripHistoryUpload -> {
+            if (compileAndUploadTripHistory(tripHistoryUpload.uploadDate)) {
+                // Update the status to 'completed' if successfully compiled and uploaded.
+                tripHistoryUpload.status = TripHistoryUploadStatus.COMPLETED.getValue();
+                Persistence.tripHistoryUploads.replace(tripHistoryUpload.id, tripHistoryUpload);
+            }
+        });
+    }
+
+    /**
+     * Obtain anonymize trip data for the given date, write to zip file, upload the zip file to S3 and finally delete
+     * the zip file from local disk.
+     */
+    private static boolean compileAndUploadTripHistory(Date date) {
+        Date startOfDay = getStartOfDay(date);
+        String zipFileName = getFileName(startOfDay, ZIP_FILE_NAME_SUFFIX);
+        File tempZipFile = null;
+        try {
+            String tempFile = String.join("/", FileUtils.getTempDirectory().getAbsolutePath(), zipFileName);
+            FileUtils.writeFileToZip(
+                tempFile,
+                getFileName(startOfDay, DATA_FILE_NAME_SUFFIX),
+                getAnonymizedTripHistory(startOfDay, getEndOfDay(date))
```

Comment:
I'm a little worried that this might create a massive JSON file that would result in an out-of-memory error. Is there a way you can write this data in a streaming fashion?

Reply:
Done.